Implemented a "sonata_block_exists" twig function

mweimerskirch opened the pull request:
This adds a "sonata_block_exists" twig function that checks if a given block type exists.

I also added "exists" methods to ``BlockHelper``, ``BlockHelperInterface``, ``BlockContextManager`` and ``BlockContextManagerInterface`` as well as ``BlockLoaderChain``, ``ServiceLoader`` and ``BlockLoaderInterface``. The latter 3 however already have a method called "support" which from the name sounds like it should do something similar, but it seems to always just return true. I didn't find out what that "support" method is really used for, so any help in understanding what it does is appreciated.

It is needed for https://github.com/sonata-project/SonataAdminBundle/pull/3469, but should be useful for other use cases as well.

OskarStark commented on Block/BlockContextManagerInterface.php:
```diff
@@ -39,6 +39,13 @@ public function addSettingsByType($type, array $settings, $replace = false);
+     *
+     * @return bool
+     */
+    public function exists($type);
```
adding a method here is BC break

we can merge this without this method and add a separate PR for the next major IMO

ping @Soullivaneuh

mweimerskirch replied:
IMHO it's not a good idea to remove the method from the interface. Yes, it is a BC break, but given that the method is called from within the bundle itself (from the Twig Extension), it should be marked as required in all implementations. But that's just my opinion.

dbu replied:
adding a method to an interface is a BC break, and according to semver, this is not OK within minor versions. if i wrote my own implementation of this interface and update a minor version, i would suddenly see errors.

the BC way of doing this would be an additional interface that extends this interface to add the exists method, and the code that wants to do the exists checks would check if that additional interface is present. in the next major version, you could merge the two interfaces and simplify the code.

mweimerskirch replied:
@dbu Ok, I can do that. What should I call it though? "BlockContextManagerExtendedInterface", "BlockContextManagerNewInterface", ...?

rande replied:
@mweimerskirch just remove the function from the interface from now. It can be added in the next major version.

OskarStark commented on Block/BlockLoaderInterface.php:
```diff
@@ -16,6 +16,13 @@
+     *
+     * @return bool
+     */
+    public function exists($type);
```
same here

OskarStark commented:
I like this PR

ping @sonata-project/contributors

mweimerskirch commented:
I removed the methods from the two interfaces and squashed the commits

rande commented:
@mweimerskirch There is one CS to fix and we are ready to merge ;) Can you add a bit of documentation too ?

Thanks

mweimerskirch commented:
Should be ok now.

rande commented:
@mweimerskirch I think you php-cs-fixer is not up to date: https://travis-ci.org/sonata-project/SonataBlockBundle/jobs/95455163

There are more CS errors now :/

Soullivaneuh commented on Block/Loader/ServiceLoader.php:
```diff
@@ -30,6 +30,18 @@ public function __construct(array $types)
+     */
+    public function exists($type)
+    {
+        return in_array($type, $this->types);
```
strict comparison.

mweimerskirch commented:
Sorry, I changed the docblock at the last minute and forgot the trailing "."
Feeling a bit stupid now ;-)
I also added the strict check @Soullivaneuh suggested.

Soullivaneuh commented on Twig/Extension/BlockExtension.php:
```diff
@@ -36,6 +36,9 @@ public function __construct(BlockHelper $blockHelper)
     public function getFunctions()
     {
         return array(
+            new \Twig_SimpleFunction('sonata_block_exists',
```
Please follow PSR-2 standard: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md#46-method-and-function-calls

Soullivaneuh commented:
Too late. But this would be great to be corrected. ;-)

dbu commented:
erm. but won't we now get fatal errors when somebody is using a custom block type and the twig function is used? i would check method_exists and throw an exception.

also, there should be an issue for the milestone of the next major version to adjust the interface and remove the method_exists checks.

WouterJ commented:
@dbu's comment still applies, applications will no have hard errors when the block loader or block context manager doesn't have an `exists()` method (e.g. when using the Symfony CMF).

Also, instead of allowing the `$block` argument passed to `sonata_block_render`, this only allows passing the `$type`. This means it'll break with loaders using other block info, like the CMF loader.

At last, this is missing tests and docs. Making it hard to find out how to use it and how the loader should behave. What should the behaviour of this method be when the passed block info is not supported?
